Add WTF and bmalloc modulemaps

adetaylor opened the pull request:
#### 2169be6bc35c6343ad968b04f1896016a2ae0805
<pre>
Add WTF and bmalloc modulemaps
<a href="https://bugs.webkit.org/show_bug.cgi?id=292791">https://bugs.webkit.org/show_bug.cgi?id=292791</a>
<a href="https://rdar.apple.com/151023888">rdar://151023888</a>

Reviewed by Elliott Williams.

This commit adds modulemap files for WTF and bmalloc. These are necessary
because Swift/C++ interop interprets C++ headers using the &quot;-fmodules&quot; option,
interpreting the C++ codebase as clang modules.

Ultimately this PR results in just two new module.modulemap
files alongside the headers. However, this can cause build time
differences in all the transitive dependencies of WTF and bmalloc
because they&apos;ll interpret header files differently.

For now, this PR just puts such modulemaps in place for OS X 26+.
Future PRs will extend the modulemaps into other targets.

The approach taken for the two modulemaps is different.

* bmalloc has a hard-coded modulemap, defining a single monolithic module.
* WTF has an autogenerated modulemap, where each header typically
  resides in a small submodule.

The approach of many small submodules was previously used in
WebKit_Internal, and is used here because Swift/C++ interoperability otherwise
tries to interpret the whole module, and that results in errors such as:

  /usr/include/c++/v1/__memory/construct_at.h:47:58: note: candidate template ignored: substitution failure [with _Tp = WTF::TextBreakIterator, _Args = &lt;const WTF::TextBreakIterator &amp;&gt;]: call to deleted constructor of &apos;WTF::TextBreakIterator&apos;
_LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp* __construct_at(_Tp* __location, _Args&amp;&amp;... __args) {

This is presumed to be <a href="https://rdar.apple.com/148760413">rdar://148760413</a> or related overinstantiation
of templates. We&apos;ll likely need to solve those in future, but meanwhile
splitting the module into lots of submodules limits the blast radius of
such issues and allows us to make progress with Swift.

Why is one manually generated and the other autogenerated?
Per the comments in the bmalloc modulemap, we only want to expose
a hand-curated set of headers in the bmalloc module, and therefore
there&apos;s no benefit to autogenerating it. It&apos;s specifically the
dependency on a certain value of OS_UNFAIR_LOCK_INLINE which
forces this decision, and that&apos;s discussed in
<a href="https://rdar.apple.com/153287512">rdar://153287512</a>.

Whereas, for WTF we want to expose all the headers, and thus evolving the
modulemap to add new headers and remove older headers would be error-prone and
time consuming - hence we autogenerate.

For the WTF mdoulemap, we use a script and a TOML file that describes the
quirks of given header files (for example, whether they&apos;re interdependent or
can be used in plain C contexts). See the TOML files in this PR and their
comments for a description of these quirks. Thanks to Elliott Williams for
writing the script.

The script converts a tapi filelist-style header listing
into a modulemap, supporting both monolithic and submodule layouts.
This process works well for libraries like WTF which have no
umbrella header, and which have no concept of public vs private API.

We do not export the modulemaps to the build output on pre-OS X 15;
doing so confuses the build of PALSwift.

This commit also adjusts some headers to be more compatible with
modules, adding missing includes, and ensuring all includes are
outside of any extern &quot;C&quot; blocks.

This PR is the third attempt at landing modulemaps for WTF and bmalloc.
The first time didn&apos;t work because it turned out that we needed to resolve
three issues first, to unbreak downstream consumers:

* We needed to move away from reliance on UCHAR_TYPE in WTF.
  (<a href="https://rdar.apple.com/153817245">rdar://153817245</a>).
* Downstream consumers of WebKit were previously using some WTF definitions
  without including WTF headers. That ceases to work when WTF becomes a
  module; we have removed that dependency (<a href="https://rdar.apple.com/154081231">rdar://154081231</a>)
* #45625 resolves an issue with AppServerSupport which starts to appear
  once a modulemap is in place. There were also small fixes to the
  modularization of various system dependencies included within WebKit.
  (Because some of these fixes in WebKit&apos;s dependencies have only landed
  in recent SDKs, we do not export the modulemaps when older SDKs
  are in use.)

The second time didn&apos;t work because we needed to:
* Be robust to different versions of the python tomli module, some of
  which require binary input and some of which require text input.
* Not attempt to use modules in downlevel builds. (Previously we were
  checking the XCode version; we should have checked the target SDK
  version and that&apos;s what we now do).

* Source/WTF/Configurations/WTF.xcconfig:
* Source/WTF/Configurations/modulemap.toml: Added.
* Source/WTF/Scripts/modulemap-from-tapi-filelist.py: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/Compiler.h:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebKit/WebKitSwift/IdentityDocumentServices/WKIdentityDocumentPresentmentDelegate.h:
* Source/bmalloc/Configurations/bmalloc.xcconfig:
* Source/bmalloc/Configurations/module.modulemap: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:

Co-authored-by: Elliott Williams &lt;emw@apple.com&gt;
Canonical link: <a href="https://commits.webkit.org/297469@main">https://commits.webkit.org/297469@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4a1e35e47b7793f7d68bca5ba3f7f09e1e10048f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111840 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31502 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21979 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117858 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62071 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113803 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32184 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40085 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84960 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35650 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114787 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25699 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100648 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65399 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25024 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18782 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61698 "Built successfully") | 
| [❌ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/104328 "Found 1 new JSC stress test failure: stress/catch-from-function-inlining-simd-function.js.ftl-eager-no-cjit (failure)") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95079 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18854 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121131 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/110398 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38870 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28899 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93835 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39250 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96905 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93657 "Found 2 new API test failures: /WebKitGTK/TestWebProcessExtensions:/webkit/WebKitWebView/web-process-crashed, /WebKitGTK/TestWebKitPolicyClient:/webkit/WebKitPolicyClient/new-window-policy (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23909 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38834 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16622 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34889 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38767 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44262 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/134662 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38408 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36259 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41733 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40122 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

emw-apple commented on Configurations/WebKitTargetConditionals.xcconfig:
```diff
@@ -231,6 +240,14 @@ WK_MACOS_1500 = $(WK_MACOS_1500_$(WK_MACOS_BEFORE_1500));
 WK_MACOS_1500_YES = _MACOS_BEFORE_1500;
 
+WK_MACOS_1500_2600 = $(WK_MACOS_1500_2600_$(WK_MACOS_BEFORE_1500));
+WK_MACOS_1500_2600 = $(WK_MACOS_2600);
```
I wonder what's going on here… you seem to have followed the pattern correctly, but this line is just overwriting the line above it. 

I don't really know how these variants are used—I don't recall seeing them elsewhere in our xcconfigs, and we don't have equivalent definitions for embedded OSes.

adetaylor replied:
I'm not using them either, and I don't understand it. I'll keep them so as to maintain the (weird) pattern.

emw-apple commented on Source/WTF/Configurations/WTF.xcconfig:
```diff
@@ -46,3 +46,13 @@ SYSTEM_FRAMEWORK_SEARCH_PATHS = $(inherited) $(WK_PRIVATE_SDK_DIR)$(SYSTEM_LIBRA
 WK_HEADERS_VFS_FILE = $(CPP_HEADERMAP_PRODUCT_HEADERS_VFS_FILE);
+
+// A WTF modulemap confuses the building of PALSwift on older SDKs
+// For now we enable only on the most recent MacOS X SDKs and we will
```
nit:
```suggestion
// For now we enable only on the most recent macOS SDK and we will
```

emw-apple commented on Source/WTF/Scripts/modulemap-from-tapi-filelist.py:
```diff
@@ -0,0 +1,93 @@
+parser.add_argument('--relative-to', help="path to WTF in build products or the SDK")
+args = parser.parse_args()
+
+if sys.version_info < (3, 11):
+    # tomli post 1.2.0 requires 'rb';
+    # tomli pre 1.1.0 requires text.
```
FWIW, I recently ended up vendoring `tomli` into the codebase for SPI checking. So we have a stable copy at `Tools/Scripts/libraries/webkitapipy/webkitapipy/_vendor/tomli`.

We would need to do some work to get that path into the filtered sources we use for production-style building, but this could be a helpful simplification down the road.

emw-apple commented on Source/bmalloc/Configurations/bmalloc.xcconfig:
```diff
@@ -21,6 +21,17 @@
 // OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 
+// A bmalloc modulemap confuses the building of PALSwift on older SDKs
+// For now we enable only on the most recent MacOS X SDKs and we will
```
same nit as above:

```suggestion
// For now we enable only on the most recent macOS SDK and we will
```

emw-apple reviewed (approved):
Makes sense – left a couple small comments, please feel welcome to land when you're ready.

adetaylor committed:
Add WTF and bmalloc modulemaps
https://bugs.webkit.org/show_bug.cgi?id=292791
rdar://151023888

Reviewed by Elliott Williams.

This commit adds modulemap files for WTF and bmalloc. These are necessary
because Swift/C++ interop interprets C++ headers using the "-fmodules" option,
interpreting the C++ codebase as clang modules.

Ultimately this PR results in just two new module.modulemap
files alongside the headers. However, this can cause build time
differences in all the transitive dependencies of WTF and bmalloc
because they'll interpret header files differently.

For now, this PR just puts such modulemaps in place for OS X 26+.
Future PRs will extend the modulemaps into other targets.

The approach taken for the two modulemaps is different.

* bmalloc has a hard-coded modulemap, defining a single monolithic module.
* WTF has an autogenerated modulemap, where each header typically
  resides in a small submodule.

The approach of many small submodules was previously used in
WebKit_Internal, and is used here because Swift/C++ interoperability otherwise
tries to interpret the whole module, and that results in errors such as:

  /usr/include/c++/v1/__memory/construct_at.h:47:58: note: candidate template ignored: substitution failure [with _Tp = WTF::TextBreakIterator, _Args = <const WTF::TextBreakIterator &>]: call to deleted constructor of 'WTF::TextBreakIterator'
_LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX20 _Tp* __construct_at(_Tp* __location, _Args&&... __args) {

This is presumed to be rdar://148760413 or related overinstantiation
of templates. We'll likely need to solve those in future, but meanwhile
splitting the module into lots of submodules limits the blast radius of
such issues and allows us to make progress with Swift.

Why is one manually generated and the other autogenerated?
Per the comments in the bmalloc modulemap, we only want to expose
a hand-curated set of headers in the bmalloc module, and therefore
there's no benefit to autogenerating it. It's specifically the
dependency on a certain value of OS_UNFAIR_LOCK_INLINE which
forces this decision, and that's discussed in
rdar://153287512.

Whereas, for WTF we want to expose all the headers, and thus evolving the
modulemap to add new headers and remove older headers would be error-prone and
time consuming - hence we autogenerate.

For the WTF mdoulemap, we use a script and a TOML file that describes the
quirks of given header files (for example, whether they're interdependent or
can be used in plain C contexts). See the TOML files in this PR and their
comments for a description of these quirks. Thanks to Elliott Williams for
writing the script.

The script converts a tapi filelist-style header listing
into a modulemap, supporting both monolithic and submodule layouts.
This process works well for libraries like WTF which have no
umbrella header, and which have no concept of public vs private API.

We do not export the modulemaps to the build output on pre-OS X 15;
doing so confuses the build of PALSwift.

This commit also adjusts some headers to be more compatible with
modules, adding missing includes, and ensuring all includes are
outside of any extern "C" blocks.

This PR is the third attempt at landing modulemaps for WTF and bmalloc.
The first time didn't work because it turned out that we needed to resolve
three issues first, to unbreak downstream consumers:

* We needed to move away from reliance on UCHAR_TYPE in WTF.
  (rdar://153817245).
* Downstream consumers of WebKit were previously using some WTF definitions
  without including WTF headers. That ceases to work when WTF becomes a
  module; we have removed that dependency (rdar://154081231)
* #45625 resolves an issue with AppServerSupport which starts to appear
  once a modulemap is in place. There were also small fixes to the
  modularization of various system dependencies included within WebKit.
  (Because some of these fixes in WebKit's dependencies have only landed
  in recent SDKs, we do not export the modulemaps when older SDKs
  are in use.)

The second time didn't work because we needed to:
* Be robust to different versions of the python tomli module, some of
  which require binary input and some of which require text input.
* Not attempt to use modules in downlevel builds. (Previously we were
  checking the XCode version; we should have checked the target SDK
  version and that's what we now do).

* Source/WTF/Configurations/WTF.xcconfig:
* Source/WTF/Configurations/modulemap.toml: Added.
* Source/WTF/Scripts/modulemap-from-tapi-filelist.py: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/Compiler.h:
* Source/WebCore/PAL/pal/spi/cg/CoreGraphicsSPI.h:
* Source/WebKit/WebKitSwift/IdentityDocumentServices/WKIdentityDocumentPresentmentDelegate.h:
* Source/bmalloc/Configurations/bmalloc.xcconfig:
* Source/bmalloc/Configurations/module.modulemap: Added.
* Source/bmalloc/bmalloc.xcodeproj/project.pbxproj:

Co-authored-by: Elliott Williams <emw@apple.com>
Canonical link: https://commits.webkit.org/297469@main
